### PR TITLE
"Ungreedify" error handler middleware in isomorphic router

### PR DESCRIPTION
**Pull request by scinos**
## Background

In #12863 we introduced the capability of having error handlers for the isomorphic router. These handlers are designed to capture errors form all routes. Because their position in the middleware chain, they are the first error handler, meaning it will capture all errors thrown by any Express middleware, Express handler or any isomorphic middleware.

It makes them too generic, and as they don't forward the error, it blocks the capability of having generic error handlers in Express.

The way they work is by adding another `route` call with the error handler:

```js
export default function( router ) {
  router( '/my-url', myHandler );
  router( errorHandler );
}
```

## Changes

This PR reworks the idea of error handlers for the isomorphic router. Instead of being a standalone handler, now they are expected to come last in the chain of handlers for a route:

```js
export default function( router ) {
  router( '/my-url', myHandler, errorHandler );
}
```

Then, when any of the handlers in the chain (`myHandler` in this case) throws an error (or calls `next(error)`) the rest of the regular handlers will be skip and all error handlers in the chain will be called.

Following the previous design, these error handlers are expected to change `req.context` and the router will implicitly call the server rendering function at the end. That part has not changed.

### Preparation

Add `console.log( 'notFoundError called' );` to the function `notFoundError` in `client/my-sites/theme/controller.jsx`. We'll use this log line to verify the function is not called.

Build all required artifacts by running these commands. You only need to run them once.

```
yarn distclean
yarn
yarn build-static
yarn build-client-evergreen
NODE_ENV=production ENABLE_FEATURES=use-translation-chunks yarn build-server
```

Start the server with `DISABLE_FEATURES=use-translation-chunks yarn start-build-do`. Despite what the log says,
the server will be available in http://calypso.localhost:3000/

When you need to restart the server to pick new changes, run `NODE_ENV=production yarn build-server && DISABLE_FEATURES=use-translation-chunks yarn start-build-do`. It should only take a few seconds.

For the following tests, you should use an incognito window and disable JavaScript (as they are focused on the SSR output)

### See the error

This test should be done in master. It will show what the problems are.

* Edit the function `setupLoggedInContext` in `client/server/pages/index.js` and make it throw an error
* Change `renderServerError` in `client/server/pages/index.js` to always log the error

Restart the server and go to http://calypso.localhost:3000/log-in. You'll see the following **wrong** behaviour:

* The log in the server log is about `Cannot read property 'store' of undefined`, unrelated to the fake error you added.
* `notFoundError` is called, but the error has nothing to do with themes.

#### Isomorphic pages

Verify the following pages are SSRed correctly

http://calypso.localhost:3000/theme/seedlet
http://calypso.localhost:3000/log-in
http://calypso.localhost:3000/themes

#### Invalid theme

Go to http://calypso.localhost:3000/theme/asdf

* You should see the `Looking for great WordPress designs?` error page
* The server logs should show something like this. Ensure the message, status and level fields are the same:

	```
	{"name":"calypso","hostname":"sergio.lan","pid":59280,"level":40,"status":404,"message":"Theme Not    Found","themeSlug":"asdf","msg":"","time":"2020-10-09T10:46:46.415Z","v":0}
	```
* `notFoundError` is called

#### Express middleware error

Edit the function `setupLoggedInContext` in `client/server/pages/index.js` and make it throw an error. Restart the server and go to:

* http://calypso.localhost:3000/log-in
* http://calypso.localhost:3000/themes
* http://calypso.localhost:3000/theme/asdf

Verify these are true for both pages:

* You should get a generic "We're sorry, but an unexpected error has occurred" error page
* The server logs shows an error with `level:50` and the stack trace.
* Function `notFoundError` is not called.

#### Isomorphic middleware error

Edit the function `fetchThemeDetailsData` in `client/my-sites/theme/controller.jsx` and make it throw an error. Restart the server and go to http://calypso.localhost:3000/theme/seedlet

Verify these are true:

* You should get the `Looking for great WordPress designs?` error page.
* The server logs should show an error with `level:40` and the stack trace.
* Function `notFoundError` is called.

### CODE REVIEW

File: client/my-sites/theme/index.node.js [hunk 1]
```diff
@@ -2,8 +2,8 @@
  * Internal dependencies
  */
 
-import config from 'config';
-import { makeLayout } from 'controller';
+import config from 'calypso/config';
```
Comment (scinos) [inline client/my-sites/theme/index.node.js:5]: These import changes are unrelated and made by `eslint` autofixing rules.

**Review by tyxla — APPROVED** [review]
This tests great, code makes sense to me, and I really appreciate the suggested changes include some good documentation that makes our SSR mechanisms easier to understand. 

I'd love it if @ockham could have a look here, as I recall he had a very good understanding of this bit of Calypso.

File: client/server/isomorphic-routing/index.js [hunk 2]
```diff
@@ -74,23 +83,38 @@ function getEnhancedContext( req, res ) {
...
+			}
+		} catch ( error ) {
+			next( error );
+		}
+
+		// At this point we are in either of these scenarios:
+		// * There is an error but this middlware is not an error handler
+		// * There is not an error but this middleware is an error handler
+		// In both cases this middleware shouldn't run, so we just skip to the next one.
+		next( err );
```
Comment (ockham) [inline client/server/isomorphic-routing/index.js:109]: What if we caught an `error` through `catch`? Will both `next( error )` and `next( err )` run? Is that desired?

File: docs/isomorphic-routing.md [hunk 3]
```diff
@@ -67,14 +67,12 @@ export function notFoundError( err, context, next ) {
...
-  router( '/themes/:slug/:section?/:site_id?', details, makeLayout );
-  router( themeNotFound );
+  router( '/themes/:slug/:section?/:site_id?', details, makeLayout, themeNotFound );
 }
 ```
 
-Note that in `notFoundError`, `err` is passed as an argument to `next`. This is how error middleware chains skip regular middlewares. The rendering middleware that is implicitly called on the server after all other middlewares are invoked uses `err.status` to set the HTTP error status.
-
-On the other hand, an error-handling middleware like `themeNotFound` will be called if any other middleware before it calls `next` with an error object:
+Note that you can have multiple error-handling middlewares in your route defintion. When any of the regular middlewares throw an error (or call `next(err)`), only error-handling will be called from that point. This is how error middleware chains skip regular middlewares. The endering middleware that is implicitly called on the server after all other middlewares are invoked uses `err.status` to set the HTTP error status. It will also log an error in the server log, using
```
Comment (ockham) [inline docs/isomorphic-routing.md:74]:
```suggestion
Note that you can have multiple error-handling middlewares in your route defintion. When any of the regular middlewares throw an error (or call `next(err)`), only error-handling will be called from that point. This is how error middleware chains skip regular middlewares. The rendering middleware that is implicitly called on the server after all other middlewares are invoked uses `err.status` to set the HTTP error status. It will also log an error in the server log, using
```

File: docs/isomorphic-routing.md [hunk 4]
```diff
@@ -67,14 +67,12 @@ export function notFoundError( err, context, next ) {
...
+severity `error` if status is >= 500, `info` otherwhise.
```
Comment (ockham) [inline docs/isomorphic-routing.md:75]: I thought

```suggestion
severity `error` if status is >= 500, `warn` otherwhise.
```

? See https://github.com/Automattic/wp-calypso/pull/46307/files#diff-7213eb8b2b566b024eff874ee238cd8baab74fe13fc0eb28c9d1c946526fc8e6R37.

File: client/server/isomorphic-routing/index.js (hunk 2 above)
Comment (scinos) [reply client/server/isomorphic-routing/index.js:109]: No, that's a bug, good catch! Only one `next(err)` should be called.

**Comment by ockham** [comment]
This is looking great, @scinos! I left a few notes -- it's been a while since I last worked on this, I hope they're any useful :sweat_smile: 

Two minor concerns:
- I hope that putting the error handling middleware last doesn't rely too much on argument position for this API. The iso-routing already has a few fairly specific and implicit requirements (such as expecting a rendered layout in `context.layout` if memory serves), and we'd be adding one more. (Not that I have a much better solution to suggest...) One risk I see is that it becomes harder to turn a client-side only middleware into an isomorphic one. (OTOH, since routing errors/HTTP statuses are kind of a non-entity in client-side routing, it's probably fine to require a different concept here.)
- The `combineMiddlewares`/`applyMiddleware` stuff still reads a bit confusing. Not sure there's much we can do, I guess it's inherent in its recursive nature. 

No blockers, and probably fine to ignore. Just wanted to have it on track in case we want to discuss more/re-evaluate later.
